Route.kt

```diff
@@ -0,0 +1,509 @@
+package com.mapzen.valhalla
+
+import com.mapzen.helpers.GeometryHelper.getBearing
+import com.mapzen.model.ValhallaLocation
+import org.json.JSONArray
+import org.json.JSONObject
+import java.lang.Math.toRadians
+import java.util.ArrayList
+import java.util.HashSet
+
+open class Route {
+
+    companion object {
+        const val KEY_TRIP = "trip"
+        const val KEY_LEGS = "legs"
+        const val KEY_SHAPE = "shape"
+        const val KEY_MANEUVERS = "maneuvers"
+        const val KEY_UNITS = "units"
+        const val KEY_LENGTH = "length"
+        const val KEY_STATUS = "status"
+        const val KEY_TIME = "time"
+        const val KEY_LOCATIONS = "locations"
+        const val KEY_SUMMARY = "summary"
+        const val CLOSE_TO_DESTINATION_THRESHOLD_METERS: Int = 20
+        const val CLOSE_TO_NEXT_LEG_THRESHOLD_METERS: Int = 5
+        const val LOST_THRESHOLD_METERS: Int = 50
+        const val CORRECTION_THRESHOLD_METERS: Int = 1000
+        const val CLOCKWISE_DEGREES: Double = 90.0
+        const val COUNTERCLOCKWISE_DEGREES: Double = -90.0
+        const val REVERSE_DEGREES: Int = 180
+        const val LOCATION_FUZZY_EQUAL_THRESHOLD_DEGREES: Double = 0.00001
+    }
+
+    lateinit var rawRoute: JSONObject
+    /**
+     * Because https://valhalla.mapzen.com/route does not use http status codes, "status" key
+     * in response can indicate too many requests in which case no poly line will be present
+     */
+    private var poly: ArrayList<Node>? = null
+    /**
+     * Because https://valhalla.mapzen.com/route does not use http status codes, "status" key
+     * in response can indicate too many requests in which case no instructions will be present
+     */
+    private var instructions: ArrayList<Instruction>? = null
+    var units: Router.DistanceUnits = Router.DistanceUnits.KILOMETERS
+    var currentLeg: Int = 0
+    private val seenInstructions = HashSet<Instruction>()
+    private var lost: Boolean = false
+    /**
+     * Snapped location along route poly line
+     */
+    private var lastFixedLocation: ValhallaLocation? = null
+    private var currentInstructionIndex: Int = 0
+    var totalDistanceTravelled: Double = 0.0
+    private var beginningRouteLostThresholdMeters: Int? = null
+
+    constructor(jsonString: String) {
+        setJsonObject(JSONObject(jsonString))
+    }
+
+    constructor(jsonObject: JSONObject) {
+        setJsonObject(jsonObject)
+    }
+
+    fun setJsonObject(jsonObject: JSONObject) {
+        this.rawRoute = jsonObject
+        if (foundRoute()) {
+            initializeDistanceUnits(jsonObject)
+            initializePolyline(jsonObject.getJSONObject(KEY_TRIP).getJSONArray(KEY_LEGS).
+                    getJSONObject(0).getString(KEY_SHAPE))
+            initializeTurnByTurn(jsonObject.getJSONObject(KEY_TRIP).getJSONArray(KEY_LEGS).
+                    getJSONObject(0).getJSONArray(KEY_MANEUVERS))
```

Comment:
Ok! This looks much better :)

No need to make another pull request but can you push a commit that removes the extra copies of the source files from the project root.

Also there are now some compile errors in the unit tests that need to be resolved due to changing `locations` from an array to an `List`.

https://circleci.com/gh/mapzen/on-the-road_android/348

```
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:140: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[0].lat).doesNotContain("1.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:141: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[0].lat).contains("3.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:142: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[1].lat).contains("5.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:163: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[0].lat).contains("1.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:164: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[0].lon).contains("2.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:165: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[1].lat).contains("3.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac                                 ^
> Building 85% > :library:compileDebugUnitTestJavaWithJavac/home/ubuntu/on-the-road_android/library/src/test/java/com/mapzen/valhalla/RouterTest.java:166: error: array required, but List<Location> found
> Building 85% > :library:compileDebugUnitTestJavaWithJavac        assertThat(json.locations[1].lon).contains("4.0");
> Building 85% > :library:compileDebugUnitTestJavaWithJavac  
```